### PR TITLE
UI: Panel

### PR DESCRIPTION
**Pull request by Amstutz**
Proposal for UI Panel Components, to be discussed in the next JF.

### CODE REVIEW

File: src/UI/Component/Panel/Factory.php [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+	 *      Cards: >
+	 *        Often Cards are used in Decks to display multiple uniformly structured chunks of Data horizontally and vertically.
+	 *
+	 * rules:
+	 *   usage:
+	 *      1: In Forms Standard Panels MUST be used to group different sections into Form Parts.
+	 *      2: Standard Panels SHOULD be used in the Center Content as primary Container for grouping content of varying content.
+	 * ---
+	 * @param string $title
+	 * @param mixed $content \ILIAS\UI\Component\Component[] | \ILIAS\UI\Component\Component
```
Comment (klees) [inline src/UI/Component/Panel/Factory.php]: I think the CLASS1|CLASS2 notation could be used instead of mixed, so you could pull the part in the back to the front. You could also `use \ILIAS\UI\Component\Component` to get `Component[]|Component`.

File: src/UI/Component/Panel/Factory.php [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	 *      Standard Panel: >
+	 *        The Standard Panel might contain a Sub Panel.
+	 *      Card: >
+	 *        The Sub Panels may contain one card.
+	 *
+	 * rules:
+	 *   usage:
+	 *      1: Sub Panels MUST only be inside Standard Panels
+	 *   composition:
+	 *      1: Sub Panels MUST NOT contain Sub Panels or Standard Panels as content.
```
Comment (klees) [inline src/UI/Component/Panel/Factory.php:55]: This can be checked easily in your implementation of `Panel\Sub`.

File: src/UI/Component/Panel/Factory.php [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	 *       filled with user-generated content harvested by that very structure (i.e. participants’ answers to
+	 *       the question, self-evaluation of competence).
+	 *   composition: >
+	 *       They are composed of a Standard Panel which contains several Block Panels. They might also contain
+	 *       a card to display information meta information in their first block.
+	 *   effect: >
+	 *       Report Panels are predominantly used for displaying data. They may however comprise links or buttons.
+	 *   rivals:
+	 *      Standard Panels: >
+	 *        The Report Panels contains one Standard Panel containing the block panels used to structure information.
```
Comment (klees) [inline src/UI/Component/Panel/Factory.php]: Is this correct? If i understand correctly, this conflicts with the actual parameters to report, which takes Subpanels.

File: src/UI/Component/Panel/Factory.php [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	 *
+	 * rules:
+	 *   usage:
+	 *      1: >
+	 *         Report Panels SHOULD be used when user generated content of two sources (i.e results, guidelines in a template)
+	 *         is to be displayed alongside each other.
+	 *   interaction:
+	 *      1: Links MAY open new views.
+	 *      2: Buttons MAY trigger actions or inline editing.
```
Comment (klees) [inline src/UI/Component/Panel/Factory.php:88]: Do these to lines really belong here?

File: src/UI/Component/Panel/Panel.php [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+/**
+ * This describes how a panel could be modified during construction of UI.
+ */
+interface Panel extends \ILIAS\UI\Component\Component {
+	/**
+	 * @param string $title Title of the Panel
+	 * @return \ILIAS\UI\Component\Panel\Panel
+	 */
+	public function withTitle($title);
```
Comment (klees) [inline src/UI/Component/Panel/Panel.php]: Do we really want to change the title after construction?

File: src/UI/Component/Panel/Report.php [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+/**
+ * This describes how a Report could be modified during construction of UI.
+ */
+interface Report extends Panel {
+	/**
+	 * @param string $title Title of the Report
+	 * @return \ILIAS\UI\Component\Panel\Report
+	 */
+	public function withTitle($title);
```
Comment (klees) [inline src/UI/Component/Panel/Report.php]: No need to duplicate this, its in Panel already.

File: src/UI/Component/Panel/Report.php [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+	/**
+	 * @return string $title Title of the Report
+	 */
+	public function getTitle();
```
Comment (klees) [inline src/UI/Component/Panel/Report.php]: No need to duplicate this, its in Panel already.

File: src/UI/Component/Panel/Factory.php (hunk 4 above)
Comment (Amstutz) [inline src/UI/Component/Panel/Factory.php:88]: We talked about MAY rules before. They are of limited use and might just be part of the description. Point here is, that we like to make clear, for what Buttons and Links might be used in Reports. Do you think, they should be kicked out here?

File: src/UI/Component/Panel/Report.php [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+	/**
+	 * @param \ILIAS\UI\Component\Panel\Sub[] $sub_panels Sub Panels used to structure the report.
+	 * @return \ILIAS\UI\Component\Panel\Report
+	 */
+	public function withSubPanels($sub_panels);
```
Comment (klees) [inline src/UI/Component/Panel/Report.php]: Do we really want to change this after construction?

File: src/UI/Component/Panel/Factory.php (hunk 4 above)
Comment (klees) [inline src/UI/Component/Panel/Factory.php:88]: No, this actually looked like a Copy-Paste-Error to me, as i didn't understand what these have to do with report panels =)

File: src/UI/Component/Panel/Factory.php (hunk 3 above)
Comment (Amstutz) [inline src/UI/Component/Panel/Factory.php]: Legacy naming, good catch!

File: src/UI/Component/Panel/Standard.php [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+/**
+ * This describes a Standard Panel.
+ */
+interface Standard extends Panel {
+	/**
+	 * @param mixed $content \ILIAS\UI\Component\Component[] | \ILIAS\UI\Component\Component
+	 * @return \ILIAS\UI\Component\Panel\Standard
+	 */
+	public function withContent($content);
```
Comment (klees) [inline src/UI/Component/Panel/Standard.php]: Do we really want to change this after construction?

File: src/UI/Component/Panel/Sub.php [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * This describes a Sub Panel.
+ */
+interface Sub extends Panel {
+	/**
+	 * @param mixed $content \ILIAS\UI\Component\Component[] | \ILIAS\UI\Component\Component
+	 * @return \ILIAS\UI\Component\Panel\Sub
+	 */
+	public function withContent($content);
```
Comment (klees) [inline src/UI/Component/Panel/Sub.php]: Do we really want to change this after construction?

File: src/UI/Component/Panel/Sub.php [hunk 11]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+/* Copyright (c) 2016 Timon Amstutz <timon.amstutz@ilub.unibe.ch> Extended GPL, see docs/LICENSE */
+
+namespace ILIAS\UI\Component\Panel;
+
+/**
+ * This describes a Sub Panel.
+ */
+interface Sub extends Panel {
```
Comment (klees) [inline src/UI/Component/Panel/Sub.php:10]: You could introduce another interface here that bundles the content-part of Sub and Standard.

File: src/UI/Component/Panel/Factory.php (hunk 1 above)
Comment (klees) [inline src/UI/Component/Panel/Factory.php]: Same in other places, btw.

File: src/UI/Factory.php [hunk 12]
```diff
@@ -253,4 +253,22 @@ public function listing();
...
+	 *   purpose: >
+	 *     Panels are used to group titled content.
+	 *   composition: >
+	 *      Panels consist of a header and content section. They form one Gestalt and so build a perceivable
+	 *      cluster of information.
+	 *   effect: The effect of interaction with panels heavily depends on their content.
+	 *
+	 * rules:
+	 *   wording:
+	 *      1: Panels MUST contain a heading.
```
Comment (klees) [inline src/UI/Factory.php]: Maybe change to "Panels MUST have a title." as this is the parameter we actually supply to the panel constructors?

File: src/UI/Implementation/Component/Panel/Factory.php [hunk 13]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+/* Copyright (c) 2016 Timon Amstutz <timon.amstutz@ilub.unibe.ch> Extended GPL, see docs/LICENSE */
+
+namespace ILIAS\UI\Implementation\Component\Panel;
+
+use ILIAS\UI\Component\Panel as P;
+
+/**
+ * Class Factory
```
Comment (klees) [inline src/UI/Implementation/Component/Panel/Factory.php:10]: Just for curiosity: Is this a line PHPStorm always creates? I stumbled upon it very often already, as i consider this non-informative.

File: src/UI/Implementation/Component/Panel/Panel.php [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+	 * @inheritdoc
+	 */
+	public function getTitle() {
+		return $this->title;
+	}
+
+	/**
+	 * @inheritdoc
+	 */
+	public function withContent($content){
```
Comment (klees) [inline src/UI/Implementation/Component/Panel/Panel.php]: Implementing this here is not a real problem, but this also hints at the fact, that there might be another interface for Panels with any content.

File: src/UI/Component/Panel/Sub.php (hunk 11 above)
Comment (Amstutz) [inline src/UI/Component/Panel/Sub.php:10]: Is the rule each component should represent an interface only one way or two way? From your comment I gather it is one way. What name would you chose for this interface?

File: tests/UI/Component/Panel/PanelTest.php [hunk 15]
```diff
@@ -0,0 +1,241 @@
...
+	public function test_render_report() {
+		$f = $this->getFactory();
+		$fp = $this->getPanelFactory();
+		$r = $this->getDefaultRenderer();
+		$sub = $fp->sub("Title",array());
+		$card = $f->card("Card Title");
+		$sub = $sub->withCard($card);
+
+		$report = $fp->report("Title",$sub);
+		$html = new DOMDocument();
```
Comment (klees) [inline tests/UI/Component/Panel/PanelTest.php]: This pattern looks nice. Maybe we could abstract it a little and promote it to the base class for tests?

Comment (klees) [inline tests/UI/Component/Panel/PanelTest.php]: Like "assertHTMLEquals" or something similar?

File: src/UI/Implementation/Component/Panel/Factory.php (hunk 13 above)
Comment (Amstutz) [inline src/UI/Implementation/Component/Panel/Factory.php:10]: :-) 

It depends on how one works. In my case PHP Storms complains if a class does not provide a description by Doc Comment. An easy fix for this complaint is to auto generate one. This is kind of a mental hack since the problem is not visible anymore. You are obviously right by pointing out that the issue of not having a description is actually still there.

File: src/UI/Component/Panel/Sub.php (hunk 11 above)
Comment (klees) [inline src/UI/Component/Panel/Sub.php:10]: I'm not sure, i understand what you are asking. You could introduce `interface PanelWithContent` (proper naming pending) that has the `*Content` methods. This would deduplicate the `*Content` methods in Sub and Panel.

File: tests/UI/Component/Panel/PanelTest.php (hunk 15 above)
Comment (Amstutz) [inline tests/UI/Component/Panel/PanelTest.php]: I stumbled upon multiple issues while testing rather nested html output from our beloved template engine. Wrapping it with a DOMDocument helped reducing some of the problems.

+1 for assertHTMLEquals

Comment (klees) [inline tests/UI/Component/Panel/PanelTest.php]: Who is going to implement it, me or you, or...?

**Comment by klees** [comment]
Will have another look on monday, not focused enough atm.

File: src/UI/Component/Panel/Sub.php (hunk 11 above)
Comment (Amstutz) [inline src/UI/Component/Panel/Sub.php:10]: Yes I could. What I am asking is: Do we need a Component named "Panel->WithContent" if we introduce this interface. I hope not.

File: tests/UI/Component/Panel/PanelTest.php (hunk 15 above)
Comment (Amstutz) [inline tests/UI/Component/Panel/PanelTest.php]: Whoever blinks first. 
...
Oke, ill do it.

File: src/UI/Component/Panel/Sub.php (hunk 11 above)
Comment (lechimp-p) [inline src/UI/Component/Panel/Sub.php:10]: I don't think we need that component.

File: src/UI/Component/Panel/Factory.php [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+	 *       filled with user-generated content harvested by that very structure (i.e. participants’ answers to
+	 *       the question, self-evaluation of competence).
+	 *   composition: >
+	 *       They are composed of a Standard Panel which contains several Block Panels. They might also contain
+	 *       a card to display information meta information in their first block.
+	 *   effect: >
+	 *       Report Panels are predominantly used for displaying data. They may however comprise links or buttons.
+	 *   rivals:
+	 *      Standard Panels: >
+	 *        The Report Panels contains one Standard Panel containing the sub panels used to structure information.
```
Comment (lechimp-p) [inline src/UI/Component/Panel/Factory.php]: This seems to contradict the signature of the factory method. The sig states, that the report panel contains sub panels.

File: src/UI/Component/Panel/Factory.php [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+	 *         is to be displayed alongside each other.
+	 *   interaction:
+	 *      1: Links MAY open new views.
+	 *      2: Buttons MAY trigger actions or inline editing.
+	 * ---
+	 * @param string $title
+	 * @param \ILIAS\UI\Component\Panel\Sub[] $sub_panels
+	 * @return \ILIAS\UI\Component\Panel\Report
+	 */
+	public function report($title,$sub_panels);
```
Comment (klees) [inline src/UI/Component/Panel/Factory.php:94]: Typehint on `array $sub_panels` possible.

File: src/UI/Implementation/Component/Panel/Panel.php [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+class Panel implements C\Panel\Panel {
+	use ComponentHelper;
+
+	/**
+	 * @var string
+	 */
+	protected  $title;
+
+	/**
+	 * @var mixed content \ILIAS\UI\Component\Component[] | \ILIAS\UI\Component\Component
```
Comment (klees) [inline src/UI/Implementation/Component/Panel/Panel.php]: This is always Component[] by __construct.

File: src/UI/Implementation/Component/Panel/Panel.php [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+	 */
+	private  $content;
+
+
+	/**
+	 * @param string $title
+	 * @param mixed $content \ILIAS\UI\Component\Component[] | \ILIAS\UI\Component\Component
```
Comment (klees) [inline src/UI/Implementation/Component/Panel/Panel.php]: `Component[]|Component`

File: src/UI/Implementation/Component/Panel/Renderer.php [hunk 20]
```diff
@@ -0,0 +1,121 @@
...
+	}
+
+	/**
+	 * @param Component\Panel\Standard $component
+	 * @param RendererInterface $default_renderer
+	 * @return string
+	 */
+	protected function render_standard(Component\Panel\Standard $component, RendererInterface $default_renderer)
+	{
+		$tpl = $this->getTemplate("tpl.panel.html", true, true);
```
Comment (klees) [inline src/UI/Implementation/Component/Panel/Renderer.php]: Use a separate template for standard panel and report panel please. This would allow our designers to change them independently.

File: src/UI/Implementation/Component/Panel/Report.php [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+
+	/**
+	 * @param string $title
+	 * @param \ILIAS\UI\Component\Panel\Sub[]
+	 */
+	public function __construct($title,$sub_panels) {
+		$this->checkStringArg("title",$title);
+		$sub_panels= $this->toArray($sub_panels);
+		$types = [C\Panel\Sub::class];
+		$this->checkArgListElements("content", $sub_panels, $types);
```
Comment (klees) [inline src/UI/Implementation/Component/Panel/Report.php]: This checks the `sub_panels` argument, not `content`.

File: src/UI/Implementation/Component/Panel/Report.php [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+	/**
+	 * @var \ILIAS\UI\Component\Panel\Sub[]
+	 */
+	private  $sub_panels;
+
+	/**
+	 * @param string $title
+	 * @param \ILIAS\UI\Component\Panel\Sub[]
+	 */
+	public function __construct($title,$sub_panels) {
```
Comment (klees) [inline src/UI/Implementation/Component/Panel/Report.php]: The `Report` class extends `Panel`, but doesn't invoke its constructor and thus leaves the inherited property `content` uninitialized, which makes `Panel::getContent` promise something it can't fullfill. Overall it looks to me as if `Report` is basically `Panel` with an additional constraints on the content. Maybe we should model it just like that.

Comment (Amstutz) [inline src/UI/Implementation/Component/Panel/Report.php]: Currently we are kind of sitting between fragmented proposals. It would be helpful for me if you look first at the complete picture before proposing single fragments. Otherwise it is hard for me to work with your suggestions. I end up writing improvements for interface/templates/implementation/tests/examples 3 times.

My current solution is not good (on a structural level even buggy), however it is somewhat a result of fragmented proposals. WithContent was moved around quite a bit during the discussion.

From your current solution I gather that you propose the following:
We have an one interface panel. From my understanding of your proposal, the withTitle and withContent part. All panels would inherit from a base class panel implementing this interface, so no extra interface would be needed, right?

--> I can live with that, note however that if done as proposed, we can not type hint anymore on Panel\Sub in the withContent function of panel.

File: src/UI/Component/Panel/Sub.php [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+	 * @return mixed content \ILIAS\UI\Component\Component[] | \ILIAS\UI\Component\Component
+	 */
+	public function getContent();
+
+	/**
+	 * Sets the card to be displayed on the right of the Sub Panel
+	 * @param \ILIAS\UI\Component\Card\Card $card
+	 * @return Sub
+	 */
+	public function withCard($card);
```
Comment (klees) [inline src/UI/Component/Panel/Sub.php]: Typehint on `Card` please.

File: src/UI/Implementation/Component/Panel/Sub.php [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+	 * Card to be displayed on the right of the Sub Panel
+	 * @var \ILIAS\UI\Component\Card\Card
+	 */
+	private $card = null;
+
+	/**
+	 * @inheritdoc
+	 */
+	public function withCard($card){
+		$this->checkArgInstanceOf("Card",$card,C\Card\Card::class);
```
Comment (klees) [inline src/UI/Implementation/Component/Panel/Sub.php]: If we typehint we don't need that call to `checkArgInstanceOf`.

File: src/UI/Implementation/Component/Panel/Sub.php [hunk 25]
```diff
@@ -0,0 +1,40 @@
...
+/**
+ * Class Sub
+ * @package ILIAS\UI\Implementation\Component\Sub
+ */
+class Sub extends Panel implements C\Panel\Sub {
+	use ComponentHelper;
+
+	/**
+	 * Card to be displayed on the right of the Sub Panel
+	 * @var \ILIAS\UI\Component\Card\Card
```
Comment (klees) [inline src/UI/Implementation/Component/Panel/Sub.php]: `Card|null` please.

**Comment by klees** [comment]
Maybe we should introduce scrutinizer for the src-directory to help us with reviews.

File: tests/UI/Base.php [hunk 26]
```diff
@@ -92,4 +93,24 @@ public function getDefaultRenderer() {
...
+	public function assertHTMLEquals($expected_html_as_string,$html_as_string){
+		$html = new DOMDocument();
+		$html->formatOutput = true;
+		$html->preserveWhiteSpace = false;
+
+		$expected = new DOMDocument();
+		$expected->formatOutput = true;
+		$expected->preserveWhiteSpace = false;
+
+		$html->loadXML($this->normalizeHTML($html_as_string));
```
Comment (klees) [inline tests/UI/Base.php]: Do we still need my homemade `normalizeHTML` or does `DOMDocument` do something similar for us?

File: tests/UI/Component/Listing/ListingTest.php [hunk 27]
```diff
@@ -103,26 +103,6 @@ public function test_render_ordered_listing() {
...
-	public function test_descriptive_invalid_items1() {
-		$f = $this->getListingFactory();
-
-		try{
-			$f->descriptive("1","1");
-		}catch(Exception $e){
-			$this->assertEquals($e->getCode(), E_RECOVERABLE_ERROR);
-		}
-	}
-
```
Comment (klees) [inline tests/UI/Component/Listing/ListingTest.php:125]: Why did you remove that?

Comment (Amstutz) [inline tests/UI/Component/Listing/ListingTest.php:125]: Due to your recommendations we type hint on array for lists here. This results in an error in PHP7 (maybe also older versions) if string is given due to the type hint on PHP level. I did regard this test as superfluous since it makes limited sense to tests php type hinting capability. 

In this commit this is mainly here, that the tests run again.

Comment (klees) [inline tests/UI/Component/Listing/ListingTest.php:125]: Oka!

**Comment by matthiaskunkel** [comment]
Jour Fixe, August 15, 2016: We highly appreciate this PR and please Timon to merge it into trunk.

File: src/UI/Implementation/Component/Panel/Renderer.php (hunk 20 above)
Comment (Amstutz) [inline src/UI/Implementation/Component/Panel/Renderer.php]: The essence of a report panel is, that it consists of a panel and subpanel. I think it to be redundant here to generate an exact copy of the template just for the reason your proposing. In the tech meeting on thursday, we agreed that we should not set dom elements classes with variables. however we should also look to not generate too much redundancy in templates as alex pointed out.

It is possible that requirements for panels change in future, if so, we can think about splitting the templates.

File: tests/UI/Base.php (hunk 26 above)
Comment (Amstutz) [inline tests/UI/Base.php]: Currently we do to remove the new line breaks.

File: src/UI/Component/Panel/Factory.php (hunk 16 above)
Comment (Amstutz) [inline src/UI/Component/Panel/Factory.php]: I slightly changed the description.

File: src/UI/Implementation/Component/Panel/Panel.php (hunk 18 above)
Comment (Amstutz) [inline src/UI/Implementation/Component/Panel/Panel.php]: No it is not, it is Component[]|Component
